File: Cube.cs
using System;

class Program
{
    static float A, B, C;

    static float cubeWidth = 20;
    static int width = 160, height = 44;
    static float[] zBuffer = new float[160 * 44];
    static char[] buffer = new char[160 * 44];
    static int backgroundASCIICode = '.';
    static int distanceFromCam = 100;
    static float horizontalOffset;
    static float K1 = 40;
    static float incrementSpeed = 0.6f;

    static float x, y, z;
    static float ooz;
    static int xp, yp;
    static int idx;

    static void CalculateForSurface(float cubeX, float cubeY, float cubeZ, char ch)
    {
        x = CalculateX(cubeX, cubeY, cubeZ);
        y = CalculateY(cubeX, cubeY, cubeZ);
        z = CalculateZ(cubeX, cubeY, cubeZ) + distanceFromCam;

        ooz = 1 / z;

        xp = (int)(width / 2 + horizontalOffset + K1 * ooz * x * 2);
        yp = (int)(height / 2 + K1 * ooz * y);

        idx = xp + yp * width;
        if (idx >= 0 && idx < width * height)
        {
            if (ooz > zBuffer[idx])
            {
                zBuffer[idx] = ooz;
                buffer[idx] = ch;
            }
        }
    }

    static float CalculateX(float i, float j, float k)
    {
        return j * MathF.Sin(A) * MathF.Sin(B) * MathF.Cos(C) - k * MathF.Cos(A) * MathF.Sin(B) * MathF.Cos(C) +
               j * MathF.Cos(A) * MathF.Sin(C) + k * MathF.Sin(A) * MathF.Sin(C) + i * MathF.Cos(B) * MathF.Cos(C);
    }

    static float CalculateY(float i, float j, float k)
    {
        return j * MathF.Cos(A) * MathF.Cos(C) + k * MathF.Sin(A) * MathF.Cos(C) -
               j * MathF.Sin(A) * MathF.Sin(B) * MathF.Sin(C) + k * MathF.Cos(A) * MathF.Sin(B) * MathF.Sin(C) -
               i * MathF.Cos(B) * MathF.Sin(C);
    }

    static float CalculateZ(float i, float j, float k)
    {
        return k * MathF.Cos(A) * MathF.Cos(B) - j * MathF.Sin(A) * MathF.Cos(B) + i * MathF.Sin(B);
    }

    static void Main()
    {
        Console.Clear();
        while (true)
        {
            Array.Fill(buffer, (char)backgroundASCIICode);
            Array.Clear(zBuffer, 0, width * height);
            cubeWidth = 20;
            horizontalOffset = -2 * cubeWidth;

            // first cube
            for (float cubeX = -cubeWidth; cubeX < cubeWidth; cubeX += incrementSpeed)
            {
                for (float cubeY = -cubeWidth; cubeY < cubeWidth; cubeY += incrementSpeed)
                {
                    CalculateForSurface(cubeX, cubeY, -cubeWidth, '@');
                    CalculateForSurface(cubeWidth, cubeY, cubeX, '$');
                    CalculateForSurface(-cubeWidth, cubeY, -cubeX, '~');
                    CalculateForSurface(-cubeX, cubeY, cubeWidth, '#');
                    CalculateForSurface(cubeX, -cubeWidth, -cubeY, ';');
                    CalculateForSurface(cubeX, cubeWidth, cubeY, '+');
                }
            }

            cubeWidth = 10;
            horizontalOffset = 1 * cubeWidth;

            // second cube
            for (float cubeX = -cubeWidth; cubeX < cubeWidth; cubeX += incrementSpeed)
            {
                for (float cubeY = -cubeWidth; cubeY < cubeWidth; cubeY += incrementSpeed)
                {
                    CalculateForSurface(cubeX, cubeY, -cubeWidth, '@');
                    CalculateForSurface(cubeWidth, cubeY, cubeX, '$');
                    CalculateForSurface(-cubeWidth, cubeY, -cubeX, '~');
                    CalculateForSurface(-cubeX, cubeY, cubeWidth, '#');
                    CalculateForSurface(cubeX, -cubeWidth, -cubeY, ';');
                    CalculateForSurface(cubeX, cubeWidth, cubeY, '+');
                }
            }

            cubeWidth = 5;
            horizontalOffset = 8 * cubeWidth;

            // third cube
            for (float cubeX = -cubeWidth; cubeX < cubeWidth; cubeX += incrementSpeed)
            {
                for (float cubeY = -cubeWidth; cubeY < cubeWidth; cubeY += incrementSpeed)
                {
                    CalculateForSurface(cubeX, cubeY, -cubeWidth, '@');
                    CalculateForSurface(cubeWidth, cubeY, cubeX, '$');
                    CalculateForSurface(-cubeWidth, cubeY, -cubeX, '~');
                    CalculateForSurface(-cubeX, cubeY, cubeWidth, '#');
                    CalculateForSurface(cubeX, -cubeWidth, -cubeY, ';');
                    CalculateForSurface(cubeX, cubeWidth, cubeY, '+');
                }
            }

            Console.SetCursorPosition(0, 0);
            for (int k = 0; k < width * height; k++)
            {
                Console.Write(k % width == 0 ? Environment.NewLine : buffer[k].ToString());
            }

            A += 0.05f;
            B += 0.05f;
            C += 0.01f;
            System.Threading.Thread.Sleep(50);
        }
    }
}

File: Cube.fs
open System

let mutable A, B, C = 0.0, 0.0, 0.0

let mutable cubeWidth = 20.0
let width, height = 160, 44
let zBuffer = Array.create (width * height) 0.0
let buffer = Array.create (width * height) ' '
let backgroundASCIICode = '.'
let distanceFromCam = 100.0
let mutable horizontalOffset = 0.0
let K1 = 40.0
let incrementSpeed = 0.6

let mutable x, y, z = 0.0, 0.0, 0.0
let mutable ooz = 0.0
let mutable xp, yp, idx = 0, 0, 0

let calculateX i j k =
    j * (sin A) * (sin B) * (cos C) - k * (cos A) * (sin B) * (cos C) +
    j * (cos A) * (sin C) + k * (sin A) * (sin C) + i * (cos B) * (cos C)

let calculateY i j k =
    j * (cos A) * (cos C) + k * (sin A) * (cos C) -
    j * (sin A) * (sin B) * (sin C) + k * (cos A) * (sin B) * (sin C) -
    i * (cos B) * (sin C)

let calculateZ i j k =
    k * (cos A) * (cos B) - j * (sin A) * (cos B) + i * (sin B)

let calculateForSurface cubeX cubeY cubeZ ch =
    x <- calculateX cubeX cubeY cubeZ
    y <- calculateY cubeX cubeY cubeZ
    z <- calculateZ cubeX cubeY cubeZ + distanceFromCam

    ooz <- 1.0 / z

    xp <- int (float width / 2.0 + horizontalOffset + K1 * ooz * x * 2.0)
    yp <- int (float height / 2.0 + K1 * ooz * y)

    idx <- xp + yp * width
    if idx >= 0 && idx < width * height then
        if ooz > zBuffer.[idx] then
            zBuffer.[idx] <- ooz
            buffer.[idx] <- ch

[<EntryPoint>]
let main argv =
    Console.Clear()

    while true do
        buffer.[0 .. width * height - 1] <- Array.init (width * height) (fun _ -> backgroundASCIICode)
        Array.fill zBuffer 0 (width * height) 0.0
        cubeWidth <- 20.0
        horizontalOffset <- -2.0 * cubeWidth

        for cubeX in -cubeWidth .. incrementSpeed .. cubeWidth do
            for cubeY in -cubeWidth .. incrementSpeed .. cubeWidth do
                calculateForSurface cubeX cubeY -cubeWidth '@'
                calculateForSurface cubeWidth cubeY cubeX '$'
                calculateForSurface -cubeWidth cubeY -cubeX '~'
                calculateForSurface -cubeX cubeY cubeWidth '#'
                calculateForSurface cubeX -cubeWidth -cubeY ';'
                calculateForSurface cubeX cubeWidth cubeY '+'

        cubeWidth <- 10.0
        horizontalOffset <- 1.0 * cubeWidth

        for cubeX in -cubeWidth .. incrementSpeed .. cubeWidth do
            for cubeY in -cubeWidth .. incrementSpeed .. cubeWidth do
                calculateForSurface cubeX cubeY -cubeWidth '@'
                calculateForSurface cubeWidth cubeY cubeX '$'
                calculateForSurface -cubeWidth cubeY -cubeX '~'
                calculateForSurface -cubeX cubeY cubeWidth '#'
                calculateForSurface cubeX -cubeWidth -cubeY ';'
                calculateForSurface cubeX cubeWidth cubeY '+'

        cubeWidth <- 5.0
        horizontalOffset <- 8.0 * cubeWidth

        for cubeX in -cubeWidth .. incrementSpeed .. cubeWidth do
            for cubeY in -cubeWidth .. incrementSpeed .. cubeWidth do
                calculateForSurface cubeX cubeY -cubeWidth '@'
                calculateForSurface cubeWidth cubeY cubeX '$'
                calculateForSurface -cubeWidth cubeY -cubeX '~'
                calculateForSurface -cubeX cubeY cubeWidth '#'
                calculateForSurface cubeX -cubeWidth -cubeY ';'
                calculateForSurface cubeX cubeWidth cubeY '+'

        Console.SetCursorPosition(0, 0)
        for k in 0 .. width * height - 1 do
            match k % width with
            | 0 -> Console.WriteLine()
            | _ -> Console.Write(buffer.[k])


        A <- A + 0.05
        B <- B + 0.05
        C <- C + 0.01
        System.Threading.Thread.Sleep(10)

    0 
